doc/man3/X509_LOOKUP_meth_new.pod

```diff
@@ -150,10 +150,16 @@ the X509_LOOKUP context, the type of the X509_OBJECT being requested, parameters
 related to the lookup, and an X509_OBJECT that will receive the requested
 object.
 
-Implementations should use either X509_OBJECT_set1_X509() or
-X509_OBJECT_set1_X509_CRL() to set the result. Any method data that was
-created as a result of the new_item function set by
-X509_LOOKUP_meth_set_new_item() can be accessed with
+Implementations must add objects it finds to the B<X509_STORE> object
+using X509_STORE_add_cert() or X509_STORE_add_crl().  This "duplicates"
+the object by incrementing its reference counter.
+
+Implementations should also use either X509_OBJECT_set1_X509() or
+X509_OBJECT_set1_X509_CRL() to set the result.  This also "duplicates"
+the object by incrementing its reference counter.
```

Comment:
But this still results in a mem leak IIUC because X509_STORE_CTX_get_by_subject ups the ref count a third time:

https://github.com/openssl/openssl/blob/ad7e17dd6c8a3931da0fa9a06e80cf498278ef27/crypto/x509/x509_lu.c#L307-L323

Reply:
Yes, but you can "free" obj->data.x509 / obj->data.crl after setting it with `X509_OBJECT_set1_*`, that will reduce the refcount and avoid that leak. Have a look at my latest changes in #8442, I have made a point of using documented methods entirely.

Reply:
Yes....but the docs don't suggest you need to do this!!

Reply:
Yeah, but that's a separate problem from this immediate one, by much of the `X509_STORE_CTX` functionality still being undocumented...  so I'm unsure how much caveat I want to dive into in this PR, as it seems we need to make a larger effort documenting this stuff.

Anyway, if you do have an idea for some simple wording to add here, I'm all ears.

Reply:
> Yeah, but that's a separate problem from this immediate one,

I saw this issue as the primary issue in need of clarification in the docs, and the whole point!

"Implementations should also use either X509_OBJECT_set1_X509() or X509_OBJECT_set1_X509_CRL() to set the result.  This also "duplicates" the object by incrementing its reference counter. However the X509_STORE_CTX_get_by_subject() function also increases the reference count which leads to one too many references being held. Therefore applications should additionally call X509_free() or X509_CRL_free() to decrement the reference count again"

An alternative is to introduce the functions X509_OBJECT_set0_X509() and X509_OBJECT_set0_X509_CRL() and recommend that applications use those instead.

Reply:
> An alternative is to introduce the functions X509_OBJECT_set0_X509() and X509_OBJECT_set0_X509_CRL() and recommend that applications use those instead.

But that doesn't really solve the problem with existing stuff.

I'll add the text that you wrote, thank you

Reply:
> But that doesn't really solve the problem with existing stuff.

It's not a silver bullet for all problems. But it makes writing the code a bit neater (prevents calling a set1 and then immediately freeing the ref).

Reply:
Sure.  But that's out of scope for this PR